### PR TITLE
feat: sticky workers

### PR DESCRIPTION
**Pull request by grutt**
# Description

Often it's desirable to run certain steps on workers in a certain state (i.e. memory config, loaded models, or assigned to the same worker throughout). This PR adds flexible support to define worker state as `key:value` pairs and specify desired state in the step definition.

For further detail, see discussion #375 

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## What's Changed

- [x] Adds WorkerLabel table for the current worker state
- [x] Adds StepDesiredWorkerLabel for the desired state of a worker for a step
- [x] Revise AcquireWorkerSemaphore 
- [x] Update relevant step run events to include desired and actual label state and render in UI
- [x] Add StickyState
- [x] Add "sticky" option to workflow runs to add future desired state to include first step worker id

## TODO
- [ ] Squash migrations
- [ ] Docs
- [ ] Py SDK
- [ ] Go SDK

Related: https://github.com/hatchet-dev/hatchet-typescript/pull/213

### CODE REVIEW

File: api-contracts/openapi/components/schemas/metadata.yaml [hunk 1]
```diff
@@ -105,9 +105,8 @@ APIResourceMeta:
       type: string
       description: "the id of this resource, in UUID format"
       example: bb214807-246e-43a5-a25d-41761d1cff9e
-      minLength: 36
+      minLength: 0
```
Comment (grutt) [inline api-contracts/openapi/components/schemas/metadata.yaml:108]: I'm not sure how we want to handle this, I chose bigint for labels but still want to use tables...

File: api-contracts/openapi/components/schemas/worker.yaml [hunk 2]
```diff
@@ -1,3 +1,18 @@
...
+      $ref: "./metadata.yaml#/APIResourceMeta"
+    key:
+      type: string
+      description: The key of the label.
+    value:
+      type: string
+      description: The value of the label.
+  required:
+    - metadata
+    - key
```
Comment (abelanger5) [inline api-contracts/openapi/components/schemas/worker.yaml:13]: Is `value` not required?

File: frontend/app/src/pages/main/workers/$worker/index.tsx [hunk 3]
```diff
@@ -204,6 +204,28 @@ export default function ExpandedWorkflowRun() {
...
+        </h3>
+        <div className="mb-4 text-sm">
+          Worker labels are key-value pairs that can be used to prioritize
+          assignment of steps to specific workers.
+        </div>
+        <div className="flex gap-2">
+          {!worker.labels || worker.labels.length === 0 ? (
+            <>
+              <>No Labels Assigned. [Docs]</>
+              {/* TODO add doc link */}
```
Comment (abelanger5) [inline frontend/app/src/pages/main/workers/$worker/index.tsx]: Address todo?

File: internal/services/dispatcher/server.go [hunk 4]
```diff
@@ -166,6 +173,51 @@ func (s *DispatcherImpl) Register(ctx context.Context, request *contracts.Worker
...
+func (s *DispatcherImpl) UpsertWorkerLabels(ctx context.Context, request *contracts.UpsertWorkerLabelsRequest) (*contracts.UpsertWorkerLabelsResponse, error) {
+	tenant := ctx.Value("tenant").(*dbsqlc.Tenant)
+
+	_, err := s.upsertLabels(ctx, sqlchelpers.UUIDFromStr(request.WorkerId), request.Labels)
+
+	if err != nil {
+		return nil, err
+	}
+
+	s.l.Error().Msgf("Received upsert worker affinities request for worker %s", request.WorkerId)
```
Comment (abelanger5) [inline internal/services/dispatcher/server.go]: Remove error line or change to debug?

File: pkg/repository/prisma/dbsqlc/step_runs.sql [hunk 5]
```diff
@@ -500,14 +523,95 @@ locked_step_runs AS (
...
+        "intValue",
+        "required",
+        "weight",
+        "comparator"
+    FROM
+        "StepDesiredWorkerLabel"
+    WHERE
+        "stepId" = (SELECT "stepId" FROM locked_step_runs)
+),
+evaluated_affinities AS (
```
Comment (abelanger5) [inline pkg/repository/prisma/dbsqlc/step_runs.sql:553]: Seems that we're calling this CTE even when there isn't a desired worker label in the step run, which will increase latency - can we case this CTE on there being a desired label?

File: api/v1/server/oas/transformers/worker.go [hunk 6]
```diff
@@ -10,6 +11,32 @@ import (
 	"github.com/hatchet-dev/hatchet/pkg/repository/prisma/dbsqlc"
 )
 
+func ToWorkerLabels(labels []*dbsqlc.ListWorkerLabelsRow) *[]gen.WorkerLabel {
+	resp := make([]gen.WorkerLabel, len(labels))
+
+	for i := range labels {
+
+		var value *string
+		if v, ok := labels[i].Value.(string); ok {
```
Comment (abelanger5) [inline api/v1/server/oas/transformers/worker.go]: Can we use `.Value.Valid` and `.Value.String` here instead of casting to string and *string?

**Review by abelanger5 — CHANGES_REQUESTED** [review]

File: api-contracts/openapi/components/schemas/worker.yaml (hunk 2 above)
Comment (grutt) [reply api-contracts/openapi/components/schemas/worker.yaml:13]: no, undefined will effectively unset the label

**Review by abelanger5 — APPROVED** [review]

**Pull request merged**